docs/tutorials/sparse/train.md

```diff
@@ -0,0 +1,256 @@
+```
+
+### Storage Type Fallback
+
+For operators that don't specialize in certain sparse arrays, you can still use them with sparse inputs with some performance penalty. In MXNet, dense operators require all inputs and outputs to be in the dense format. If sparse inputs are provided, MXNet will convert sparse inputs into dense ones temporarily so that the dense operator can be used. If sparse outputs are provided, MXNet will convert the dense outputs generated by the dense operator into the provided sparse format. Warning messages will be printed when such a storage fallback event happens.
+
+
+```python
+# `log` operator doesn't support sparse inputs at all, but we can fallback on the dense implementation
+csr_log = mx.sym.log(a)
+# `elemwise_add` operator doesn't support adding csr with row_sparse, but we can fallback on the dense implementation
+csr_rsp_add = mx.sym.elemwise_add(b, c)
+fallback_exec = mx.sym.Group([csr_rsp_add, csr_log]).simple_bind(ctx=mx.cpu(), a=shape, b=shape, c=shape)
+fallback_exec.forward()
+fallback_add = fallback_exec.outputs[0]
+fallback_log = fallback_exec.outputs[1]
+{'fallback_add': fallback_add, 'fallback_log': fallback_log}
+```
+
+### Inspecting Storage Types of the Symbol Graph (Work in Progress)
+
+When the environment variable `MXNET_EXEC_LOG_LEVEL` is set to `INFO`, MXNet will log the information of the graph executor,
+including storage types of operators' inputs and outputs in the graph. For example, we can inspect the storage types of
+a linear classification network with sparse operators as follows:
+
+
+```python
+# Set logging level for executor
+import mxnet as mx
+import os
+os.environ['MXNET_EXEC_LOG_LEVEL'] = "INFO"
+# Data in csr format
+data = mx.sym.var('data', stype='csr', shape=(32, 10000))
+# Weight in row_sparse format
+weight = mx.sym.var('weight', stype='row_sparse', shape=(10000, 2))
+bias = mx.symbol.Variable("bias", shape=(2,))
+dot = mx.symbol.sparse.dot(data, weight)
+pred = mx.symbol.broadcast_add(dot, bias)
+y = mx.symbol.Variable("label")
+output = mx.symbol.SoftmaxOutput(data=pred, label=y, name="output")
+executor = output.simple_bind(ctx=mx.cpu())
+```
+
+## Training with Module APIs
+
+In the following section we'll walk through how one can implement **linear regression** using sparse symbols and sparse optimizers.
+
+The function you will explore is: *y = x<sub>1</sub>  +  2x<sub>2</sub> + ... 100x<sub>100*, where *(x<sub>1</sub>,x<sub>2</sub>, ..., x<sub>100</sub>)* are input features and *y* is the corresponding label.
+
+### Preparing the Data
+
+In MXNet, both [mx.io.LibSVMIter](https://mxnet.incubator.apache.org/versions/master/api/python/io.html#mxnet.io.LibSVMIter)
+and [mx.io.NDArrayIter](https://mxnet.incubator.apache.org/versions/master/api/python/io.html#mxnet.io.NDArrayIter)
+support loading sparse data in CSR format. In this example, we'll use the `NDArrayIter`.
+
+You may see some warnings from SciPy. You don't need to worry about those for this example.
```

Comment:
Hmm, I don't find use of SciPy below. What is the reason to see possible warnings from SciPy?

Reply:
mx.test_utils.rand_ndarray may call scipy.random if scipy is available. The random function in scipy converts coo to csr. For some version of scipy, its `random` function is using a deprecated argument when performing the conversion. This is not what we can control

Reply:
This is the output you get from running the block below. 
```
/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/scipy/sparse/coo.py:200: VisibleDeprecationWarning: `rank` is deprecated; use the `ndim` attribute or function instead. To find the rank of a matrix see `numpy.linalg.matrix_rank`.
  if np.rank(self.data) != 1 or np.rank(self.row) != 1 or np.rank(self.col) != 1:
/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/scipy/sparse/compressed.py:130: VisibleDeprecationWarning: `rank` is deprecated; use the `ndim` attribute or function instead. To find the rank of a matrix see `numpy.linalg.matrix_rank`.
  if np.rank(self.data) != 1 or np.rank(self.indices) != 1 or np.rank(self.indptr) != 1:
```